go/tools/gopackagesdriver/main.go

```diff
@@ -0,0 +1,121 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"go/types"
+	"os"
+	"os/signal"
+	"strings"
+)
+
+type driverResponse struct {
+	NotHandled bool
+
+	// Sizes, if not nil, is the types.Sizes to use when type checking.
+	Sizes *types.StdSizes
+
+	// Roots is the set of package IDs that make up the root packages.
+	// We have to encode this separately because when we encode a single package
+	// we cannot know if it is one of the roots as that requires knowledge of the
+	// graph it is part of.
+	Roots []string `json:",omitempty"`
+
+	// Packages is the full set of packages in the graph.
+	// The packages are not connected into a graph.
+	// The Imports if populated will be stubs that only have their ID set.
+	// Imports will be connected and then type and syntax information added in a
+	// later pass (see refine).
+	Packages []*FlatPackage
+}
+
+var (
+	bazelBin          = getenvDefault("GOPACKAGESDRIVER_BAZEL", "bazel")
+	workspaceRoot     = os.Getenv("BUILD_WORKSPACE_DIRECTORY")
+	targetsStr        = os.Getenv("GOPACKAGESDRIVER_BAZEL_TARGETS")
+	targetsQueryStr   = os.Getenv("GOPACKAGESDRIVER_BAZEL_QUERY")
+	targetsTagFilters = os.Getenv("GOPACKAGESDRIVER_BAZEL_TAG_FILTERS")
+)
+
+func getenvDefault(key, defaultValue string) string {
+	if v, ok := os.LookupEnv(key); ok {
+		return v
+	}
+	return defaultValue
+}
+
+func signalContext(parentCtx context.Context, signals ...os.Signal) (ctx context.Context, stop context.CancelFunc) {
+	ctx, cancel := context.WithCancel(parentCtx)
+	ch := make(chan os.Signal, 1)
+	go func() {
+		select {
+		case <-ch:
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+	signal.Notify(ch, signals...)
+
+	return ctx, cancel
+}
+
+func run() error {
+	ctx, cancel := signalContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	request, err := ReadDriverRequest(os.Stdin)
+	if err != nil {
+		return fmt.Errorf("unable to read request: %w", err)
+	}
+
+	bazel, err := NewBazel(ctx, bazelBin, workspaceRoot)
+	if err != nil {
+		return fmt.Errorf("unable to create bazel instance: %w", err)
+	}
+
+	targets := []string{}
+	if targetsStr != "" {
+		targets = strings.Split(targetsStr, " ")
```

Comment:
Maybe this should use a regex for splitting?
I accidentally put two spaces between a pair of targets and everything broke.

Reply:
good one. I have pushed a fix, which also simplifies this bit. thank you